src/write_syslog.c

```diff
@@ -0,0 +1,666 @@
+ * MessageFormat - Selects the format in which messages are sent to the broker,
+ *          human or json. Defaults to human.
+ * HostTags -When set, HostTags is added to the end of the metric.
+ *   It is intended to be used for adding additional metadata to the metric
+ * with.
+ *   Dots and whitespace are not escaped in this string.
+ *   When MessageFormat is set to "human" text should be in valid syslog
+ *   structured data format. No spaces allowed.
+ *   Example:
+ *     [ovirt engine="test"][ovirt vm="test2"]
+ *   When MessageFormat is set to "json", text should be in valid Json format.
+ *   Escaping the quotation marks is required.
+ *   Example:
+ *    HostTags "\"ovirt\": {\"engine\":\"test\",\"vm\":\"test2\"}"
+ * StoreRates false|true
+ *   If set to true, convert counter values to rates. If set to false (the
+ *   default) counter values are stored as is, as an increasing integer number.
+ */
+
+#include "collectd.h"
+
+#include "common.h"
+#include "plugin.h"
+#include "utils_cache.h"
+#include "utils_random.h"
+
+#include <netdb.h>
+
+#ifndef WS_DEFAULT_NODE
+#define WS_DEFAULT_NODE "localhost"
+#endif
+
+#ifndef WS_DEFAULT_SERVICE
+#define WS_DEFAULT_SERVICE "44514"
+#endif
+
+#ifndef WS_DEFAULT_FORMAT
+#define WS_DEFAULT_FORMAT "human"
+#endif
+
+#ifndef WS_DEFAULT_PREFIX
+#define WS_DEFAULT_PREFIX "collectd"
+#endif
+
+#ifndef WS_DEFAULT_ESCAPE
+#define WS_DEFAULT_ESCAPE '.'
+#endif
+
+/* Ethernet - (IPv6 + TCP) = 1500 - (40 + 32) = 1428 */
+#ifndef WS_SEND_BUF_SIZE
+#define WS_SEND_BUF_SIZE 1428
+#endif
+
+/*
+ * Private variables
+ */
+struct ws_callback {
+  struct addrinfo *ai;
+  cdtime_t ai_last_update;
+  int sock_fd;
+
+  char *node;
+  char *service;
+  char *host_tags;
+  char *msg_format;
+  char *metrics_prefix;
+  bool store_rates;
+  bool always_append_ds;
+
+  char send_buf[WS_SEND_BUF_SIZE];
+  size_t send_buf_free;
+  size_t send_buf_fill;
+  cdtime_t send_buf_init_time;
+
+  pthread_mutex_t send_lock;
+
+  bool connect_failed_log_enabled;
+  int connect_dns_failed_attempts_remaining;
+  cdtime_t next_random_ttl;
+};
+
+static cdtime_t resolve_interval;
+static cdtime_t resolve_jitter;
+
+/*
+ * Functions
+ */
+static void ws_reset_buffer(struct ws_callback *cb) {
+  memset(cb->send_buf, 0, sizeof(cb->send_buf));
+  cb->send_buf_free = sizeof(cb->send_buf);
+  cb->send_buf_fill = 0;
+  cb->send_buf_init_time = cdtime();
+}
+
+static int ws_send_buffer(struct ws_callback *cb) {
+  ssize_t status = 0;
+
+  status = swrite(cb->sock_fd, cb->send_buf, strlen(cb->send_buf));
+  if (status != 0) {
+    ERROR("write_syslog plugin: send failed with status %zi (%s)", status,
+          STRERRNO);
+
+    close(cb->sock_fd);
+    cb->sock_fd = -1;
+
+    return -1;
+  }
+
+  return 0;
+}
+
+/* NOTE: You must hold cb->send_lock when calling this function! */
+static int ws_flush_nolock(cdtime_t timeout, struct ws_callback *cb) {
+  int status;
```

Comment:
Please always initialize local variables. The majority of these cases will go away when the previous comment is resolved.

For the remainder: https://collectd.org/review-comments#initialize-variables

Reply:
Copied from write_tsdb